presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitSource.java

```diff
@@ -262,6 +273,89 @@ public boolean isFinished(OptionalInt bucketNumber)
                 highMemorySplitSourceCounter);
     }
 
+    public static HiveSplitSource bucketedRewindable(
+            ConnectorSession session,
+            String databaseName,
+            String tableName,
+            TupleDomain<? extends ColumnHandle> compactEffectivePredicate,
+            int maxInitialSplits,
+            DataSize maxOutstandingSplitsSize,
+            HiveSplitLoader splitLoader,
+            Executor executor,
+            CounterStat highMemorySplitSourceCounter)
+    {
+        AtomicReference<State> stateReference = new AtomicReference<>(State.initial());
+        return new HiveSplitSource(
+                session,
+                databaseName,
+                tableName,
+                compactEffectivePredicate,
+                new PerBucket()
+                {
+                    private final Map<Integer, List<InternalHiveSplit>> splits = new ConcurrentHashMap<>();
+                    private final SettableFuture<?> finished = SettableFuture.create();
+
+                    @Override
+                    public ListenableFuture<?> offer(OptionalInt bucketNumber, InternalHiveSplit connectorSplit)
+                    {
+                        checkArgument(bucketNumber.isPresent(), "bucketNumber must be present");
+                        splits.computeIfAbsent(bucketNumber.getAsInt(), ignored -> new ArrayList<>()).add(connectorSplit);
+                        // Do not block "offer" when running split discovery in bucketed mode.
+                        // A limit is enforced on estimatedSplitSizeInBytes.
+                        return immediateFuture(null);
+                    }
+
+                    @Override
+                    public ListenableFuture<List<ConnectorSplit>> borrowBatchAsync(OptionalInt bucketNumber, int maxSize, Function<List<InternalHiveSplit>, BorrowResult<InternalHiveSplit, List<ConnectorSplit>>> function)
+                    {
+                        checkArgument(bucketNumber.isPresent(), "bucketNumber must be present");
+                        if (!finished.isDone()) {
+                            return finished.transformAsync(ignored -> borrowBatchAsync(bucketNumber, maxSize, function), executor);
```

Comment:
the `transformAsync` makes it harded to reason about the code, as now the splits fetching is assynchronous. 

e.g.: when rewind is called, the feature returned by this method may not have yet been done.

I would suggest to return this feature as empty. I don't think it is really makes any sense to complicate the code to optimize this one more call to the `borrowBatchAsync`.

Reply:
I think returning an empty list might be wrong because that breaks the assumption that `borrowBatchAsync` will be blocked if the splits are not ready.

https://github.com/prestodb/presto/blob/a0e24e275e7207efe7ed8a5d62eeaf9bf47f5d53/presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitSource.java#L423

Returning an empty list might create a  `ConnectorSplitBatch` whose `noMoreSplits` is set to `true`, and then `SplitBatch` whose `lastBatch` is `true`, then

https://github.com/prestodb/presto/blob/2e199b0e64b346c6641b67b70ee442f1b0617c76/presto-main/src/main/java/com/facebook/presto/execution/scheduler/SourcePartitionedScheduler.java#L220-L235

The underlying problem is that we need to differentiate between "scheduling finished" (no more splits will be loaded/offered) and "execution finished" (no more splits will be rewinded). Let's discuss offline tomorrow.

Reply:
@shixuan-fan :  As a side note, it looks to me `splits.isEmpty()` is not required; as discussed in https://github.com/prestodb/presto/issues/12662

Reply:
Let's not remove it though :laughing:

Reply:
@arhimondr : Why not  😆

Reply:
We can, but in a separate PR =)